awscrt/http.py

```diff
@@ -0,0 +1,174 @@
+                                                                 connection._on_connection_shutdown,
+                                                                 host_name,
+                                                                 port,
+                                                                 socket_options,
+                                                                 internal_conn_options_handle)
+
+        except Exception as e:
+            future.set_exception(e)
+
+        return future
+
+    def close(self):
+        """
+        Closes the connection, if you hold a reference to this instance of HttpClientConnection, on_connection_shutdown
+        will be invoked upon completion of the connection close.
+        """
+        if self._native_handle is not None:
+            _aws_crt_python.aws_py_http_client_connection_close(self._native_handle)
+
+    def is_open(self):
+        """
+        Returns True if the connection is open and usable, False otherwise.
+        """
+        if self._native_handle is not None:
+            return _aws_crt_python.aws_py_http_client_connection_is_open(self._native_handle)
+
+        return False
+
+    def make_request(self, request):
+        """
+        Makes an Http request. When the headers from the response are received, the returned future will have a result.
+        and request.response_headers will be filled in, and request.response_code will be available.
+        After this future completes, you can get the result of request.response_completed,
+        for the remainder of the response.
+        """
+        future = Future()
+
+        def on_stream_completed(error_code):
+            if error_code == 0:
+                request.response_completed.set_result(error_code)
+            else:
+                request.response_completed.set_exception(Exception(error_code))
+
+        def on_incoming_headers_received(headers, response_code, has_body):
+            request.response_headers = headers
+            request.response_code = response_code
+            request.has_response_body = has_body
+            future.set_result(response_code)
+
+        try:
+            request._stream = _aws_crt_python.aws_py_http_client_connection_make_request(self._native_handle,
+                                                                                         request,
+                                                                                         on_stream_completed,
+                                                                                         on_incoming_headers_received)
+
+        except Exception as e:
+            future.set_exception(e)
+
+        return future
+
+
+class OutgoingHttpBodyState(IntEnum):
+    InProgress = 0
+    Done = 1
+
+
+class HttpRequest(object):
+    """
+    Represents an HttpRequest to pass to HttpClientConnection.make_request(). path_and_query is the path and query portion
```

Comment:
I don't think we need a "builder/options" class like this in Python. Python handles default-argument and named-arguments really well, so it's not a bad idea to just pile args into the make_request(...) function